### PR TITLE
SDK proto compatibility guarantees and deprecation policies documentation

### PR DESCRIPTION
**Pull request by igooch**
**What type of PR is this?**

/kind documentation

**What this PR does / Why we need it**:

As part of work on In-place Agones Upgrades #3766 we are going to explicitly guarantee certain proto versions during upgrade and this documents that contract.

**Which issue(s) this PR fixes**:

Working on #3767

**Special notes for your reviewer**:

### CODE REVIEW

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+- Alpha SDK Protos (and APIs/RPCs) are subject to change between releases.
+  - A game server binary using Alpha SDKs may not be compatible with a newer sdk-server.
+  - In Alpha, incompatible changes retaining the same SDK proto message name are allowed.
+  - When we make incompatible Alpha changes, we will document the APIs involved.
+
+## SDK Deprecation Policies
+Breaking changes will be called out in upgrade documentation, alongside the build horizon, allowing operators to plan their upgrades.
+### Stable Deprecation Policies
+A Stable API may be marked as deprecated in release X and removed from Stable in release X+10.
+TODO: Does this deprecation first need to flow through Alpha and / or Beta?
```
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: Is a change (including deprecations) required to go through Beta first before being introduced into Stable?

[the remark this answers is not in the document]
Comment (zmerlynn) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: I really don't understand how we would make the deprecation flow through? In the system we agreed on, an SDK proto is going to settle at a particular stability level, or it might be bridging two - but we're not going to re-introduce it in the other stability level(s) just to deprecate it?

**Review by markmandel — COMMENTED** [review]
Meta question -- I'm wondering if this (and all other details on in-place upgrades), should be in: https://agones.dev/site/docs/installation/upgrading/

Or that section shouid be fleshed out into separate sections, since ultimately the CUJ is "I want to update in place", not "I want to learn how the sdk-server works" ? (Please correct me if I'm wrong).

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+## SDK Server Overview
+- The SDK Server is a gRPC server. The methods for communication between the SDK Client and SDK Server are defined in the [SDK proto](https://github.com/googleforgames/agones/blob/main/proto/sdk).
+- The REST API is generated directly from the SDK .proto files.
+- All other SDK Client APIs are wrappers on top of the SDK proto definitions.
+
+{{% feature publishVersion="1.43.0" %}}
+
+{{< alert title="Note" color="info">}}
+This section is specifically about the SDK proto compatibility between Agones versions and deprecation policies.
```
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md:17]: Mostly trying to point out that this page is not explaining what the SDK Server is or does. It's explaining the version compatibility and deprecation policies.

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  - In Alpha, incompatible changes retaining the same SDK proto message name are allowed.
+  - When we make incompatible Alpha changes, we will document the APIs involved.
+
+## SDK Deprecation Policies
+Breaking changes will be called out in upgrade documentation, alongside the build horizon, allowing operators to plan their upgrades.
+### Stable Deprecation Policies
+A Stable API may be marked as deprecated in release X and removed from Stable in release X+10.
+TODO: Does this deprecation first need to flow through Alpha and / or Beta?
+### Beta Deprecation Policies
+When a feature graduates from Beta to Stable at release X the API is in both protos from release X to release X+5. The Beta API is marked as deprecated in release X and removed from Beta in release X+5.
```
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md:37]: Not sure I entirely understood this comment?

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+### Alpha Deprecation Policies
+There is no guaranteed proto compatibility between releases for Alpha SDK protos. When an Alpha API graduates to Beta the API will be deleted from the Alpha proto with no overlapping release.
+An API may be removed from the Alpha proto during any release without graduating to Beta.
+
+## SDK Server APIs and Stability Levels
+
+"Legacy" indicates that this API has been in the SDK Server in a release before we began tracking proto compatibility.
+
+{{< alert title="Note" color="info">}}
+The Actions may differ from the Client SDKs depending on how each Client SDK is implemented.
```
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md:48]: All the other tables, or just the main _index.md page?

[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: Also for some reason href links inside a Note do not work.

**Comment by igooch** [comment]
> Meta question -- I'm wondering if this (and all other details on in-place upgrades), should be in: https://agones.dev/site/docs/installation/upgrading/

We definitely want to have the in-place upgrades info there. Probably most of what's in this PR could also go in there. If we switch it over I would probably both feature-gate it to when in-place upgrades is released in Alpha, and make even more explicit from what number release the info applies.

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+weight: 1001
+description: "The SDK Server is a sidecar for a GameServer that will update the GameServer Status on SDK requests."
+---
+
+## SDK Server Overview
+- The SDK Server is a gRPC server. The methods for communication between the SDK Client and SDK Server are defined in the [SDK proto](https://github.com/googleforgames/agones/blob/main/proto/sdk).
+- The REST API is generated directly from the SDK .proto files.
+- All other SDK Client APIs are wrappers on top of the SDK proto definitions.
+
+{{% feature publishVersion="1.43.0" %}}
```
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: I moved most of the docs over to Upgrades, so I featured-gated it at 1.43 to be around the time we expect in-place upgrades to go into Alpha. Even if we don't feature-gate at 1.43 we should probably call out the exact version from which the guarantees apply.

[the remark this answers is not in the document]
Comment (zmerlynn) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: Is there a reason we can't provide these guarantees right now? I don't really think this needs to be gated.

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (zmerlynn) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: I think you resolved this in the move.

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (zmerlynn) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: Instead of..
`When a feature graduates from Beta to Stable at release X the API is in both protos from [...]`

I was thinking more something like:
`When a feature graduates from Beta to Stable at release X, the API will be present in both Beta and Stable surfaces from [...]`

File: site/content/en/docs/Installation/upgrading.md [hunk 6]
```diff
@@ -128,7 +128,63 @@ upgrades.
...
+When a feature graduates from Beta to Stable at release X the API is in SDK proto surfaces from release X to release X+5. The Beta API is marked as deprecated in release X and removed from Beta in release X+5.
+A Beta API may be marked as deprecated in release X and removed from Beta in release X+5 without the API graduating to Stable.
+### Alpha Deprecation Policies
+There is no guaranteed proto compatibility between releases for Alpha SDK protos. When an Alpha API graduates to Beta the API will be deleted from the Alpha proto with no overlapping release.
+An API may be removed from the Alpha proto during any release without graduating to Beta.
+
+## SDK Server APIs and Stability Levels
+
+"Legacy" indicates that this API has been in the SDK Server in a release before we began tracking proto compatibility. \
+The Actions may differ from the [Client SDKs]({{< relref "../guides/Client SDKs/_index.md" >}}) depending on how each Client SDK is implemented.
```
Comment (zmerlynn) [inline site/content/en/docs/Installation/upgrading.md:163]: I think you can use a `{{< relref "Client SDKs" >}}` here, see https://github.com/googleforgames/agones/blob/db847357163b7ca974553bdf536a2541624ff12b/site/content/en/docs/Advanced/system-diagram.md?plain=1#L49 as an example. (I just noticed how long that line is, didn't even notice in VSCode, woops.)

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (zmerlynn) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md]: I think linking to Client SDKs is sufficient, yeah.

**Review by zmerlynn — COMMENTED** [review]
LGTM - would like @markmandel to give it a pass, though.

File: site/content/en/docs/Installation/upgrading.md [hunk 7]
```diff
@@ -128,7 +128,74 @@ upgrades.
...
-1. Scale your Fleets back up and/or recreate your GameServers. 
+1. Scale your Fleets back up and/or recreate your GameServers.
 1. Run any other tests to ensure the Agones installation is still working as expected.
 1. Close your maintenance window.
 7. Congratulations - you have now upgraded to a new version of Kubernetes! 👍
+
+{{% feature publishVersion="1.41.0" %}}
+## SDK Proto Compatibility Guarantees as of Agones v1.41.0
+Our [SDK Server]({{< relref "../guides/Client SDKs/sdk-server.md" >}}) compatibility contract: A
+game server binary using Beta and Stable SDKs will remain compatible with a _newer_ `sdk-server`,
```
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md:139]: And same throughout. I don't think it's important to mention the sidecar server.

File: site/content/en/docs/Installation/upgrading.md [hunk 8]
```diff
@@ -128,7 +128,74 @@ upgrades.
...
+Stable surfaces from release X to release X+5. The Beta API is marked as deprecated in release X and
+removed from Beta in release X+5.
+A Beta API may be marked as deprecated in release X and removed from Beta in release X+5 without the
+API graduating to Stable.
+### Alpha Deprecation Policies
+There is no guaranteed proto compatibility between releases for Alpha SDK protos. When an Alpha API
+graduates to Beta the API will be deleted from the Alpha proto with no overlapping release.
+An API may be removed from the Alpha proto during any release without graduating to Beta.
+
+## SDK Server APIs and Stability Levels
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:169]:
```suggestion
## SDK APIs and Stability Levels
```

File: site/content/en/docs/Installation/upgrading.md [hunk 9]
```diff
@@ -128,7 +128,74 @@ upgrades.
...
+
+"Legacy" indicates that this API has been in the SDK Server in a release before we began tracking
+proto compatibility. \
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:172]:
```suggestion
SDK compatibility. \
```

File: site/content/en/docs/Installation/upgrading.md [hunk 10]
```diff
@@ -128,7 +128,74 @@ upgrades.
...
 1. Start and complete your control plane upgrade(s).
 1. Start and complete your node upgrades.
-1. Scale your Fleets back up and/or recreate your GameServers. 
+1. Scale your Fleets back up and/or recreate your GameServers.
 1. Run any other tests to ensure the Agones installation is still working as expected.
 1. Close your maintenance window.
 7. Congratulations - you have now upgraded to a new version of Kubernetes! 👍
+
+{{% feature publishVersion="1.41.0" %}}
+## SDK Proto Compatibility Guarantees as of Agones v1.41.0
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:137]:
```suggestion
## SDK Proto Compatibility Guarantees
```

As per above, I think we need a _why_ header and description - and probably have a alert box of some kind saying something like "we're start work on in-place upgrades as of version 1.41.0, it's currently Alpha... etc etc"

WDYT?

**Review by markmandel — COMMENTED** [review]
Good start! I think if we focus on the _why_ rather than the _how_ at least to start, this will round out quite nicely.

File: site/content/en/docs/Installation/upgrading.md [hunk 11]
```diff
@@ -128,7 +128,74 @@ upgrades.
...
+within possible deprecation windows:
+- If your game server uses a non-deprecated Stable API, your binary will be compatible for 10
+releases (~1y) starting from the SDK version packaged.
+  - For example, if the game server uses non-deprecated stable APIs in the 1.40 SDK, it will be
+  compatible through the 1.50 SDK.
+  - Stable APIs will almost certainly be compatible beyond 10 releases, but 10 releases is
+  guaranteed.
+- If your game server uses a non-deprecated Beta API, your binary will be compatible for 5 releases
+(~6mo).
+- Alpha SDK Protos are subject to change between releases.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:149]:
```suggestion
- Alpha SDK APIs are subject to change between releases.
```

File: site/content/en/docs/Guides/Client SDKs/sdk-server.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
+---
+title: "Agones SDK Server"
```
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Guides/Client SDKs/sdk-server.md:2]: It's probably not necessary now that most the info is elsewhere.

File: site/content/en/docs/Installation/upgrading.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (igooch) [reply site/content/en/docs/Installation/upgrading.md:137]: This reason for "proto" is specifically SDK proto / SDK Server and does not apply to the client. The SDK client could be well out of date. For instance the C++, Rust, Unity, and Unreal clients don't have CountsAndLists at all even though it's going into Beta, and if no one updates them they may not have those features in GA. So it should be clear to the reader that this guarantee does not apply to the SDK clients. (Also the reason for using "proto" instead of "API" which is a bit more ambiguous.) If we were guaranteeing SDK clients we would need to have a separate discussion about deprecating clients if no one updates them regularly.

Comment (igooch) [reply site/content/en/docs/Installation/upgrading.md:137]: This is in place upgrades adjacent. The guarantees are still valid even if the user does not use in place upgrades, although it is necessary for in place upgrades to have a guaranteed window of compatibility.

[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: But as an end user I don't care about the proto - I want to know if I am using my SDK client, will that stay backward compatible.

Much of this is covered in this section of the design doc: https://github.com/googleforgames/agones/issues/3766#client-sdks - which for this documentation is the more relevant part.

> The SDK client could be well out of date.

Why would that mater?

> or instance the C++, Rust, Unity, and Unreal clients don't have CountsAndLists at all even though it's going into Beta, and if no one updates them they may not have those features in GA.

Not an issue for backward or forward compatability.

> If we were guaranteeing SDK clients we would need to have a separate discussion about deprecating clients if no one updates them regularly.

I'm not sure why we would do that? Assuming the proto doesn't change (which it shouldn't for stable APIs, they should just work forever -- but the linked above RFC covers this.

File: site/content/en/docs/Installation/upgrading.md [hunk 13]
```diff
@@ -128,7 +128,78 @@ upgrades.
...
 1. Start and complete your node upgrades.
-1. Scale your Fleets back up and/or recreate your GameServers. 
+1. Scale your Fleets back up and/or recreate your GameServers.
 1. Run any other tests to ensure the Agones installation is still working as expected.
 1. Close your maintenance window.
 7. Congratulations - you have now upgraded to a new version of Kubernetes! 👍
+
+{{% feature publishVersion="1.41.0" %}}
+## SDK Compatibility Guarantees
+In order to allow compatibility between game server binaries and `sdk-server`, a game server binary
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:138]: I was going back through the RFC - and this is probably a better paragraph to start with, then we can get into the details:
https://github.com/googleforgames/agones/issues/3766#client-sdks

File: site/content/en/docs/Installation/upgrading.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: > For instance if the client has an API in Beta and no one moves it to GA, the Beta will break after 5 releases, so anyone using that client can't move past the version when it's removed from Beta

I don't think it will. There implementation doesn't go away - it's still there, it will all still work.

> Beta Client SDK APIs should be considered generally stable, but we may alter the parameters/returns if we need to

Beta APIs _may_ change, but that's also the risk of using Beta APIs - but are definitely considered more stable than Alpha.

 > We foresee that updating Client SDKs in game server binaries will be yearly toil, if using Stable APIs, or semiannual toil, if using any Beta API; this toil can be as simple as verifying that there were no deprecations in the period involved. Given how stable our core APIs have been, it may be possible to go multiple years without updating the Client SDK.

I am seeing some more room here to flesh out exactly what is guaranteed with a Beta SDK though - I think there's some ambiguity here that could be more concrete for sure.

File: site/content/en/docs/Installation/upgrading.md [hunk 14]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
 1. Run any other tests to ensure the Agones installation is still working as expected.
 1. Close your maintenance window.
 7. Congratulations - you have now upgraded to a new version of Kubernetes! 👍
+
+{{% feature publishVersion="1.41.0" %}}
+## SDK Compatibility Guarantees
+
+The SDK compatibility contract aims to ensure smooth upgrades and reduce the need for frequent
+binary redeployments for game server users.
+What’s the point of having a compatibility guarantee?
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:141]:
```suggestion

What’s the point of having a compatibility guarantee?
```
Just so we get a new paragraph - otherwise it stays same line.

If you want to break to a newline in markdown, it's `   ` (which your editor may trim).

There's a few formatting bits in a few places like this. Might want to `make site-server` to give it a once over to ensure it's formatted how you want.

File: site/content/en/docs/Installation/upgrading.md [hunk 15]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+  compatible through the 1.50 SDK.
+  - Stable APIs will almost certainly be compatible beyond 10 releases, but 10 releases is
+  guaranteed.
+- If your game server uses a non-deprecated Beta API, your binary will be compatible for 5 releases
+(~6mo).- Alpha SDK APIs are subject to change between releases.
+  - A game server binary using Alpha SDKs may not be compatible with a newer sdk-server.
+  - In Alpha, incompatible changes retaining the same SDK proto message name are allowed.
+  - When we make incompatible Alpha changes, we will document the APIs involved.
+
+## SDK Deprecation Policies as of Agones v1.41.0
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:160]:
```suggestion
## SDK Deprecation Policies as of Agones
```

File: site/content/en/docs/Installation/upgrading.md [hunk 16]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+### Alpha Deprecation Policies
+
+There is no guaranteed proto compatibility between releases for Alpha SDK protos. When an Alpha API
+graduates to Beta the API will be deleted from the Alpha proto with no overlapping release.
+An API may be removed from the Alpha proto during any release without graduating to Beta.
+
+## SDK APIs and Stability Levels
+
+"Legacy" indicates that this API has been in the SDK Server in a release before we began tracking
+SDK compatibility. \
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:192]:
```suggestion
SDK compatibility with release 1.41.0. \
```

I think that's enough, so we don't need it in the header.

File: site/content/en/docs/Installation/upgrading.md [hunk 17]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
 1. Close your maintenance window.
 7. Congratulations - you have now upgraded to a new version of Kubernetes! 👍
+
+{{% feature publishVersion="1.41.0" %}}
+## SDK Compatibility Guarantees
+
+The SDK compatibility contract aims to ensure smooth upgrades and reduce the need for frequent
+binary redeployments for game server users.
+What’s the point of having a compatibility guarantee?
+Goal: Reduce user toil by minimizing the need for frequent binary redeployments due to SDK changes.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:142]: Love this!

File: site/content/en/docs/Installation/upgrading.md [hunk 18]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+game server binary using Beta and Stable SDKs will remain compatible with a _newer_ Agones Release,
+within possible deprecation windows:
+- If your game server uses a non-deprecated Stable API, your binary will be compatible for 10
+releases (~1y) starting from the SDK version packaged.
+  - For example, if the game server uses non-deprecated stable APIs in the 1.40 SDK, it will be
+  compatible through the 1.50 SDK.
+  - Stable APIs will almost certainly be compatible beyond 10 releases, but 10 releases is
+  guaranteed.
+- If your game server uses a non-deprecated Beta API, your binary will be compatible for 5 releases
+(~6mo).- Alpha SDK APIs are subject to change between releases.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:155]:
```suggestion
(~6mo).
- Alpha SDK APIs are subject to change between releases.
```

I think this is meant to be a new line?

File: site/content/en/docs/Installation/upgrading.md [hunk 19]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+binary redeployments for game server users.
+What’s the point of having a compatibility guarantee?
+Goal: Reduce user toil by minimizing the need for frequent binary redeployments due to SDK changes.
+Versioning: Clear documentation of API maturity levels and history (Stable, Beta, Alpha) with release information.
+
+Our SDK Server compatibility contract as of Agones v1.41.0: A
+game server binary using Beta and Stable SDKs will remain compatible with a _newer_ Agones Release,
+within possible deprecation windows:
+- If your game server uses a non-deprecated Stable API, your binary will be compatible for 10
+releases (~1y) starting from the SDK version packaged.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:149]:
```suggestion
releases (~1y) or more starting from the SDK version packaged.
```

WDYT? It read like it'll stop working after 1y (even though there is a line below it).

File: site/content/en/docs/Installation/upgrading.md [hunk 20]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+  - For example, if the game server uses non-deprecated stable APIs in the 1.40 SDK, it will be
+  compatible through the 1.50 SDK.
+  - Stable APIs will almost certainly be compatible beyond 10 releases, but 10 releases is
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:152]:
```suggestion
  - Stable APIs will almost certainly be compatible beyond 10 releases, as deprecation of stable APIs is rare, but 10 releases is
```

Something like that?

File: site/content/en/docs/Installation/upgrading.md [hunk 21]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+  - A game server binary using Alpha SDKs may not be compatible with a newer sdk-server.
+  - In Alpha, incompatible changes retaining the same SDK proto message name are allowed.
+  - When we make incompatible Alpha changes, we will document the APIs involved.
+
+## SDK Deprecation Policies as of Agones v1.41.0
+
+- Client SDK updates are not mandatory for game server binaries except for SDK proto deprecations or
+breaking Alpha API changes.
+
+- You can continue to use a binary compiled several months ago working, except in the case of a deprecation.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:165]: This sentence is a bit weird, can you rephrase?

File: site/content/en/docs/Installation/upgrading.md [hunk 22]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+
+- Expect to check if there are breaking changes to Stable APIs you use yearly or Beta APIs semi-annually.
+
+### Stable Deprecation Policies
+
+A Stable API may be marked as deprecated in release X and removed from Stable in release X+10.
+
+### Beta Deprecation Policies
+
+When a feature graduates from Beta to Stable at release X, the API will be present in both Beta and
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:177]:
```suggestion
When a SDK API feature graduates from Beta to Stable at release X, the API will be present in both Beta and
```

File: site/content/en/docs/Installation/upgrading.md [hunk 23]
```diff
@@ -128,7 +128,94 @@ upgrades.
...
+Stable surfaces from release X to release X+5. The Beta API is marked as deprecated in release X and
+removed from Beta in release X+5.
+A Beta API may be marked as deprecated in release X and removed from Beta in release X+5 without the
+API graduating to Stable.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:181]:
```suggestion
API graduating to Stable if it's determined that changes need to be made to the Beta API.
```

Just to give some reasoning.

**Review by markmandel — COMMENTED** [review]
Mainly nits, but I like this so much more!

We could probably add some details on the "how" at later stage, depending on if that's information people want or need.

Once nits are done, happy to merge.

File: site/content/en/docs/Installation/upgrading.md (hunk 21 above)
Comment (igooch) [reply site/content/en/docs/Installation/upgrading.md:165]: wtyt?

File: site/content/en/docs/Installation/upgrading.md [hunk 24]
```diff
@@ -128,7 +128,97 @@ upgrades.
...
+- If your game server uses a non-deprecated Stable API, your binary will be compatible for 10
+releases (~1y) or more starting from the SDK version packaged.
+  - For example, if the game server uses non-deprecated stable APIs in the 1.40 SDK, it will be
+  compatible through the 1.50 SDK.
+  - Stable APIs will almost certainly be compatible beyond 10 releases, as deprecation of stable
+  APIs is rare, but 10 releases is guaranteed.
+- If your game server uses a non-deprecated Beta API, your binary will be compatible for 5 releases
+(~6mo).
+- Alpha SDK APIs are subject to change between releases.
+  - A game server binary using Alpha SDKs may not be compatible with a newer sdk-server.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:159]:
```suggestion
  - A game server binary using Alpha SDKs may not be compatible with a newer Agones release if brewing changes have been made between releases.
```

File: site/content/en/docs/Installation/upgrading.md [hunk 25]
```diff
@@ -128,7 +128,97 @@ upgrades.
...
+  - In Alpha, incompatible changes retaining the same SDK proto message name are allowed.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:160]:
```suggestion
  - In Alpha, incompatible SDK changes retaining the same SDK proto message formats will continue to work.
```

File: site/content/en/docs/Installation/upgrading.md [hunk 26]
```diff
@@ -128,7 +128,97 @@ upgrades.
...
+  - When we make incompatible Alpha changes, we will document the APIs involved.
+
+## SDK Deprecation Policies as of Agones
+
+- Client SDK updates are not mandatory for game server binaries except for SDK proto deprecations or
+breaking Alpha API changes.
+
+- A binary compiled several months ago should continue to work, except in the case of a deprecation.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:168]: I'm leaning towards removing this line. I think it's redundant since you say basically the same thing in the line above. WDYT?

File: site/content/en/docs/Installation/upgrading.md [hunk 27]
```diff
@@ -128,7 +128,97 @@ upgrades.
...
+
+When a SDK API feature graduates from Beta to Stable at release X, the API will be present in both Beta and
+Stable surfaces from release X to release X+5. The Beta API is marked as deprecated in release X and
+removed from Beta in release X+5.
+A Beta API may be marked as deprecated in release X and removed from Beta in release X+5 without the
+API graduating to Stable if it's determined that changes need to be made to the Beta API.
+
+### Alpha Deprecation Policies
+
+There is no guaranteed proto compatibility between releases for Alpha SDK protos. When an Alpha API
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:188]:
```suggestion
There is no guaranteed compatibility between releases for Alpha SDKs. When an Alpha API
```

File: site/content/en/docs/Installation/upgrading.md [hunk 28]
```diff
@@ -128,7 +128,97 @@ upgrades.
...
+graduates to Beta the API will be deleted from the Alpha proto with no overlapping release.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:189]:
```suggestion
graduates to Beta the API will be deleted from the Alpha SDK with no overlapping release.
```

File: site/content/en/docs/Installation/upgrading.md [hunk 29]
```diff
@@ -128,7 +128,97 @@ upgrades.
...
+An API may be removed from the Alpha proto during any release without graduating to Beta.
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:190]:
```suggestion
An API may be removed from the Alpha SDK during any release without graduating to Beta.
```

**Review by markmandel — COMMENTED** [review]
I think this is it (hopefully 😆). Words are hard.

File: site/content/en/docs/Installation/upgrading.md (hunk 26 above)
Comment (igooch) [reply site/content/en/docs/Installation/upgrading.md:168]: Sounds good, less is usually more.

File: site/content/en/docs/Installation/upgrading.md (hunk 25 above)
Comment (igooch) [reply site/content/en/docs/Installation/upgrading.md:161]: This sentence has a slightly different meaning. Generally with proto messages a field is immutable -- if you want to make a change to a field you have to deprecate the field and create a new field with a different name. This is saying we allow changes to existing fields for Alpha only protos.

Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md:161]: oh yeah, I get you!

Since we say "  - A game server binary using Alpha SDKs may not be compatible with a newer Agones release if
  breaking changes have been made between releases." - maybe we drop this line. I don't think we need to document how it could break? WDYT?

Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md:161]: We can document in the release notes when breaks happen (which we say we will do).

File: site/content/en/docs/Installation/upgrading.md [hunk 30]
```diff
@@ -128,7 +128,95 @@ upgrades.
...
+Our SDK Server compatibility contract as of Agones v1.41.0: A
+game server binary using Beta and Stable SDKs will remain compatible with a _newer_ Agones Release,
+within possible deprecation windows:
+- If your game server uses a non-deprecated Stable API, your binary will be compatible for 10
+releases (~1y) or more starting from the SDK version packaged.
+  - For example, if the game server uses non-deprecated stable APIs in the 1.40 SDK, it will be
+  compatible through the 1.50 SDK.
+  - Stable APIs will almost certainly be compatible beyond 10 releases, as deprecation of stable
+  APIs is rare, but 10 releases is guaranteed.
+- If your game server uses a non-deprecated Beta API, your binary will be compatible for 5 releases
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:156]:
```suggestion
- If your game server uses a non-deprecated Beta API, your binary will be guaranteed compatible for 5 releases, but may be compatible past that point.
```

File: site/content/en/docs/Installation/upgrading.md [hunk 31]
```diff
@@ -128,7 +128,95 @@ upgrades.
...
+(~6mo).
+- Alpha SDK APIs are subject to change between releases.
+  - A game server binary using Alpha SDKs may not be compatible with a newer Agones release if
+  breaking changes have been made between releases.
+  - When we make incompatible Alpha changes, we will document the APIs involved.
+
+## SDK Deprecation Policies as of Agones
+
+- Client SDK updates are not mandatory for game server binaries except for SDK proto deprecations or
```
Comment (markmandel) [inline site/content/en/docs/Installation/upgrading.md:165]:
```suggestion
- Client SDK updates are not mandatory for game server binaries except for SDK when the underlying proto format has deprecations or
```

**Review by markmandel — APPROVED** [review]

**Comment by markmandel** [comment]
Flaky 1.29 Autopilot. Cleared the cluster and trying again.

**Comment by zmerlynn** [comment]
Looks great! Sent #3827 to fix a minor issue.
